scylla-cql/src/types/deserialize/value.rs

```diff
@@ -36,6 +36,40 @@ where
     ) -> Result<Self, DeserializationError>;
 }
 
+impl<'frame> DeserializeValue<'frame> for CqlValue {
+    fn type_check(_typ: &ColumnType) -> Result<(), TypeCheckError> {
+        // CqlValue accepts all possible CQL types
+        Ok(())
+    }
+
+    fn deserialize(
+        typ: &'frame ColumnType,
+        v: Option<FrameSlice<'frame>>,
+    ) -> Result<Self, DeserializationError> {
+        let mut val = ensure_not_null_slice::<Self>(typ, v)?;
+        let cql = deser_cql_value(typ, &mut val).map_err(|err| {
```

Comment:
Before deser refactor all deserialization was going trough `deser_cql_value` and all actual deserialization logic was there, right?
IIUC after you add `DeserializeValue` implementation the deserialization logic will be duplicated, right?
I think `deser_cql_value` should use `DeserializeValue` somehow and not duplicate logic.
Maybe it's addressed in later commits, if so feel free to ignore.

Reply:
This would also allow us to dump `GenericParseError`, right?

Reply:
> This would also allow us to dump `GenericParseError`, right?

Not really, unfortunately, as some low-level deserialization error (think `types` module) are represented using `ParseError`.

ATM `ParseError` has a variant of `DeserializationError`, and `DeserializationError` can contain `ParseError` inside... That's a mess, but an errors refactor is needed here to somehow split `ParseError` into finer-grained pieces. E.g. `LowLevelDeserializationError`, which would be returned from `types` module routines and it would fall into `DeserializationError`.

Reply:
> I think deser_cql_value should use DeserializeValue somehow and not duplicate logic.

In new two commits, I've refactored `deser_cql_value` and `deser_rows` to use the new framework, so much less duplication and better error messages.